### ISSUE TITLE
Refine "overdue" and introduce new "deactivated" subscription state

### ISSUE THREAD

**Issue by olmari** [TampereHacklab/mulysa#392]
## Overview:

Overdue state needs to allow door access, and after that there could be new "deactivated" status before suspended.
Bigger enhancement instead of just fixing overdue behaviour in: https://github.com/TampereHacklab/mulysa/issues/386, allowing more control over individual hacklab needs.

## Reasoning:

Humans are humans and sometimes forget things, there needs to exist some leeway between payment overdue and door access not working. Also mispayments can happen and sometimes clerical error from hacklab side. Also money still can take 2 banking days to arrive, and banking data is currently not anywhere near to real time either.

## Logic-flow:

- After active days ends, change status to overdue, keep allowing door access
  - Time perioid configurable
- After overdue days ends, change status to deactivated, deny door access, user can still catch up with payments without administrative intervention
  - Time perioid configurable
- After deactivated days overdue, change status to suspended, needs administrative intervention to remediate.
- Send remainders when states change (also keep the current pre-warning on active as is)

## Technical workflow:

One way:
Logic in overdue and deactivated works pretty similarly, overdue allows door acess, deactivated doesn't, comes one after another.
- Current "overdue" could be literally renamed as "deactivated" and test working (as current overdue function is actually deactivation by nature)
- Introduce new overdue which functions mostly the same as current overdue, exept allow door to open still
- Overdue time field in Member service settings needs rewording: "How many days service can be in overdue until it is moved to deactivated state"
- Needs new field inside Member service settings, "How many days service can be in deactivated state until it is moved to suspended state"
- Both overdue and deactivation time fields could be set to null/blank (not set) or 0, effectively skipping that status entirely, not everything is applicable to every hacklab and every service hacklab might offer.
- Make sure sure logic flow uses correct settings, so flow actually is correctly from active -> overdue -> deactivated -> suspended

**Comment by vranki** [comment]
Just a comment to "Overdue state needs to allow door access". This is incorrect, overdue state means the payments are overdue and door *must not* open until new payment is detected. This works as expected with current state definitions. The idea is to have buffer for import delays and warn user *before* entering overdue state. But feel free to re-define the states if needed. Just test really well as state machine can be error-prone.

**Comment by braaar** [comment]
I disagree with @vranki here. I haven't used mulysa in a real case, but I feel that if you miss your payment date, you shouldn't be cut off from the space immediately. This mirrors common practices in many areas of society. Overdue payments happen all the time for legitimate reasons, and if every organisation was to deny services immediately when a payment i missed, so many things would stop working smoothly. 

Electricity bills and housing rent are the classic examples, but there are many less critical areas where it would be asinine and destructive to draw a hard line like this.

Sure, there are some organisations that do block services immediately, such as streaming services. The important distinction between these and a hacklab is that it's quite trivial to lose access to your Netflix for a while. Losing access to a hacklab could mean missing out on social happenings or a project falling apart at a critical time. Another reason it's different is that you can get back into the streaming service instantaneously with a valid credit card. If you're standing outside the hacklab in freezing weather and mulysa says your payment is overdue there's nothing you can do since mulysa doesn't support instantaneous, automated re-activation.

**Comment by braaar** [comment]
That being said, I do think it's fine if mulysa supports the use case where missing a payment instantly locks you out.

I think it makes sense to consider very explicitly decoupling the business rules of when you get locked out from the payment status of the service. If privileges such as door access, software access, etc. are defined with separate, configurable rules we wouldn't have to poke as much around with the core logic of services and their statuses when changes are requested and new clubs adopt the system for their use case. 

Here are some examples of what such a rule could be like

> Allow door access if...
> The user is not BANNED or SUSPENDED AND
>  the user has a "key susbscription" with status ACTIVE or OVERDUE

> Allow door access if...
> The user is not BANNED or SUSPENDED AND
>  the user has a "key subscription" with status ACTIVE

> Allow access to remote 3D printing service if...
> the user is over 18 AND 
> the user has completed the 3D printing course AND 
> (the user has a "key subscription" with status ACTIVE OR 
> (the user has a "key subscription" with status OVERDUE AND
> the "key subscription" is overdue with less than 15 days))

If we adopt this paradigm, the overdue status can just mean that the payment is overdue, without implications for the real life consequences of that. Email alerts, access to systems like doors, SSH, software, permissions inside of the mulysa GUI, etc. can all use similar rules. Adding a new special case of overdue with its own logic seems like a mistake to me.

I've come to enjoy putting authorisation rules in one place, which is sometimes done with tools like [Oso](https://www.osohq.com/) or [Open Policy Agent](https://www.openpolicyagent.org/). Not saying we have to adopt one of those tools, but the principle is useful. 

The configuration should probably be a simple matter of selecting among a set of rules, such as the two variants of door access I outlined above. 

Functions like this one would then have a switch-case/if-elif statement inside them to choose which rule to use:
```py
    def has_door_access(self):
        """
        helper method for checking if the user has access to open door
        TODO: this should probably be checked by businesslogic
        """
        if not self.is_active:
            return False
        try:
            subscription = self.servicesubscription_set.get(
                service=config.DEFAULT_ACCOUNT_SERVICE
            )
            if subscription.state == ServiceSubscription.ACTIVE:
                return True
        except Exception:
            pass
        return False
```

I would prefer that we move the rules to one specific folder in the codebase so it's easy to keep track of all authorization logic. The member function `has_door_access` on user could still be useful, but it would just call a function from the authorization folder, not contain any logic of its own.
